pkg/tickers/retry.go

```diff
@@ -0,0 +1,94 @@
+package tickers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+	"github.com/smartcontractkit/ocr2keepers/pkg/util"
+)
+
+const (
+	retryDelay              = 500 * time.Millisecond
+	totalAttempt            = 3
+	attemptsCacheExpiration = 10 * retryDelay // 5 seconds
+)
+
+type retryTicker struct {
+	timeTicker
+	nextRetries     sync.Map // time.Time -> ocr2keepers.UpkeepPayload
+	payloadAttempts *util.Cache[int]
+}
+
+// Retry adds a retryable result to the retryTicker.
+func (rt *retryTicker) Retry(result ocr2keepers.CheckResult) error {
+	payload := result.Payload
+
+	if result.Retryable {
+		attemptCount, ok := rt.payloadAttempts.Get(payload.ID)
+		if !ok {
+			attemptCount = 0
+		}
+
+		if attemptCount >= totalAttempt {
+			fmt.Printf("Payload %s has already been tried %d times. Skipping.\n", payload.ID, totalAttempt)
+			return nil
+		}
+
+		attemptCount++
+		rt.payloadAttempts.Set(payload.ID, attemptCount, util.DefaultCacheExpiration)
+
+		nextRunTime := time.Now().Add(retryDelay)
+		rt.nextRetries.Store(nextRunTime, payload)
```

Comment:
I'm not sure what happens under the hood here, but in a high volume environment is whatever this time resolves to when used as a key granular or unique enough to ensure that we won't overwrite payloads in the map? I would have thought using the payload ID as the key would have been safer?

Reply:
Good point. What if the nextRuntime collide? I am not sure the likelihood of collision. I guess you are right, at a given time, there should be just one payload in cache, so payload is a unique identifier. Would like to hear thoughts from @EasterTheBunny too.